internal/config/config.go

```diff
@@ -0,0 +1,61 @@
+package config
+
+import (
+	"net/url"
+	"regexp"
+
+	"github.com/pkg/errors"
+	flag "github.com/spf13/pflag"
+	"github.com/spf13/viper"
+)
+
+func New() (*viper.Viper, error) {
+	config := viper.New()
+
+	err := config.BindPFlags(flag.CommandLine)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not bind config to CLI flags")
+	}
+
+	// try to get the "config" value from the bound "config" CLI flag
+	path := config.GetString("config")
+	if path != "" {
+		// try to manually load the configuration from the given path
+		err = loadConfigurationFromFile(config, path)
+	} else {
+		// otherwise try viper's auto-discovery
+		err = loadConfigurationAutomatically(config)
+	}
+
+	if err != nil {
+		return nil, errors.Wrap(err, "could not load configuration file")
+	}
+
+	setLogLevel(config.GetString("log-level"))
+
+	err = validateConfigValues(config)
+	if err != nil {
+		return nil, errors.Wrap(err, "invalid configuration value")
+	}
+
+	sanitizeConfigValues(config)
+
+	return config, nil
+}
+
+func validateConfigValues(config *viper.Viper) error {
+	sapControlUrl := config.GetString("sap-control-url")
+	if _, err := url.ParseRequestURI(sapControlUrl); err != nil {
+		return errors.Wrap(err, "invalid config value for sap-control-url")
+	}
+	return nil
+}
+
+func sanitizeConfigValues(config *viper.Viper) {
```

Comment:
perhaps the name of this helper function is better to be called

```sanitizeSAPControlUr```l` something like this

the current name of the function imply a not single-responsability and is to vague

Reply:
Yeah, I had originally done it how you're suggesting, but then I realized that we have to pass the whole configuration as a pointer argument and mutate it in place anyways, while a single responsibility function would have to be a pure one, getting a string in and putting another string out.
We can re-evaluate this in case we add more validators, but for the time being we only have one.

Reply:
I would imho suggest the contra-approach. I would for the time beeing use a `not-generic` function istead of generalizing it ( I guess we will never sanitize anh other urls).

I think the function name should not depend of the fact we give it a pointer as argument ( this is 2ndary).

We should rather have clearly in mind what the function does. Reading the function name, I guess it manipulate multiples stuff but in fact doesn't

Reply:
I'm speaking here only about the name

Reply:
hmm okay, fair enough! I guess YAGNI comes into place here. I'll change both names!